### PR TITLE
Add custom account API for plugins

### PR DESCRIPTION
**Pull request by danfinlay**
This PR introduces a set of changes that should allow plugins to provide new accounts to the MetaMask user, which should be selectable and usable the same way normal accounts are used today.

## Implementation

Converted "asset" controller (formerly used for custom asset support for plugins) into abstract "resource" controller, which is used to initialize a new `pluginAccountsController`, which is now also used for providing custom accounts to the system.

Introduced the new `accounts` controller, which merges this new `pluginAccountsController` with the old `keyringController`, while exposing a nearly equivalent interface to the `keyringController`, which formerly was the source of all of our accounts. This allows the UI to be redirected to this new controller while retaining previous functionality, merely routing account methods to their controlling code (either keyring or plugin).

Also adds a new method to plugins: `wallet.registerAccountMessageHandler(handler)`, where `handler` is an async function that accepts an `RPC` request object related to signing (`eth_sign`, `eth_signTransaction`, `personal_sign`), and returns the object expected from the equivalent `keyringController` method. (Needs better documentation, obviously, will come as I flesh out a sample custom account plugin). This callback must be registered for any plugin-provided accounts to actually do anything other than be selectable and revealable to sites.

Needs:
- [x] New method for registering signing/sending methods. (`wallet.registerAccountMessageHandler`).
- [x] Concatenation into all normal UI `accounts` listings, for seamless external interaction with extended accounts.
- [x] Create sample plugin in `mm-plugin` repo. https://github.com/MetaMask/mm-plugin/pull/31
- [x] To be tied into the normal UI so these custom accounts can be selected and used like any other. (this may be as simple as modifying a selector function, I really hope so).
- [x] The ability for app keys to be generated from custom accounts. Right now app keys are on a per-plugin basis, but should also be per-account, and since some accounts can be provided by plugins, we have a slight recursive app key generation issue to solve here. Can be solved after per-account app key generation is added. https://github.com/MetaMask/metamask-plugin-beta/issues/69

### CODE REVIEW

**Comment by danfinlay** [comment]
I basically got this working, but it’s blocked by the confirmation screen being heavily tied to “selected account” vs the “from” field of the transaction. Feels like something to fix upstream in the [LoginPerSite branch](https://github.com/MetaMask/metamask-extension/pull/7198) and then merge back down to the plugin branch, then to this custom accounts branch before merging this into the plugin branch.

**Comment by danfinlay** [comment]
Superceded by #80.

**Comment by danfinlay** [comment]
Oh my mistake, #80 merged into this branch, but this never made it to `develop`!

**Review by danjm — DISMISSED** [review]

File: app/scripts/controllers/accounts.js
Comment (rekmarks) [inline app/scripts/controllers/accounts.js]: Why not this?
```javascript
const initState = { accountrings: [], ...opts.initState }
```

File: app/scripts/controllers/accounts.js
Comment (rekmarks) [inline app/scripts/controllers/accounts.js]: Typo: `resourceAddres` should be `resourceAddress`

File: app/scripts/controllers/accounts.js
Comment (rekmarks) [inline app/scripts/controllers/accounts.js]: Minor performance gain: It seems that we re-instantiate every `accountring` regardless of whether it includes the account to be removed. We could add a check to prevent this, e.g. `if (newAccountsRingsAccounts.length !== accountring.accounts.length)`.

File: app/scripts/controllers/accounts.js
Comment (rekmarks) [inline app/scripts/controllers/accounts.js:133]: For future reference: we should standardize these errors (i.e. pick whatever number we want in the 4000-range) and add them to `eth-json-rpc-errors`

File: app/scripts/controllers/accounts.js
Comment (rekmarks) [inline app/scripts/controllers/accounts.js:176]: For future reference:
https://github.com/MetaMask/eth-hd-keyring/issues/15
https://github.com/MetaMask/eth-simple-keyring/issues/29

File: app/scripts/controllers/resource.js
Comment (rekmarks) [inline app/scripts/controllers/resource.js]: As in accounts controller, do we need this?

File: app/scripts/controllers/resource.js
Comment (rekmarks) [inline app/scripts/controllers/resource.js]: `resources and assets` -> `accounts and assets`?

**Review by rekmarks — APPROVED** [review]
Left a number of inline nits. Looks great overall! Critical addition to the plugin system.

**Review by rekmarks — DISMISSED** [review]
Pending changes from @danjm

File: app/scripts/controllers/accounts.js
Comment (danfinlay) [reply app/scripts/controllers/accounts.js:133]: Opened as https://github.com/MetaMask/metamask-snaps-beta/issues/119

**Pull request closed without merging**
